services/viva-ms/lambdas/createVivaCase.js

```diff
@@ -0,0 +1,211 @@
+import uuid from 'uuid';
+
+import config from '../../../config';
+import params from '../../../libs/params';
+import hash from '../../../libs/helperHashEncode';
+import * as request from '../../../libs/request';
+import { putItem } from '../../../libs/queries';
+import * as dynamoDB from '../../../libs/dynamoDb';
+
+const CASE_STATUS_OPEN_TO_APPLY = 'openToApply';
+const CASE_PROVIDER_VIVA = 'VIVA';
+const VADA_SSM_PARAMS = params.read(config.vada.envsKeyName);
+const CASE_SSM_PARAMS = params.read(config.vada.envsKeyName);
+
+export const main = async event => {
+  const { user } = event.detail;
+
+  const [applicationStatusError, applicationStatusResponse] = await to(
+    sendApplicationStatusRequest(user.personalNumber)
+  );
+  if (applicationStatusError) {
+    return console.error('(Viva-ms) Viva Application Status Error', applicationStatusError);
+  }
+
+  if (!isApplicationPeriodOpen(applicationStatusResponse)) {
+    return console.error(
+      '(Viva-ms) syncApplicationStatus',
+      'Application period is not open',
+      applicationStatusResponse
+    );
+  }
+
+  const [myPagesError, myPagesResponse] = await to(sendMyPagesReguest(user.personalNumber));
+
+  if (myPagesError) {
+    return console.error('(Viva-ms) Viva My Pages Request Error', applicationStatusError);
+  }
+
+  const vivaApplication = myPagesResponse.person.application.vivaapplication;
+
+  if (!vivaApplication || !vivaApplication.period) {
+    return console.error('(Viva-ms) Viva Application Perid not present in response, aborting');
+  }
+
+  if (!vivaApplication || !vivaApplication.workflowid) {
+    return console.error('(Viva-ms) Viva Application WorkflowId not present in response, aborting');
+  }
+
+  const casePrimaryKey = `USER#${user.personalNumber}`;
+  const [queryCasesError, queryCaseItems] = await to(
+    queryCasesWithWorkflowId(casePrimaryKey, vivaApplication.workflowid)
+  );
+  if (queryCasesError) {
+    return console.error('(Viva-ms) DynamoDb query on cases tabel failed', queryCasesError);
+  }
+
+  if (queryCaseItems.length > 0) {
+    return console.log('(Viva-ms) Case with WorkflowId already exists');
+  }
+
+  const [putItemError] = await to(
+    putRecurringVivaCase(casePrimaryKey, vivaApplication.workflowid, vivaApplication.period)
+  );
+  if (putItemError) {
+    return console.error('(viva-ms) syncApplicationStatus', putItemError);
+  }
+
+  return true;
+};
+
+async function queryCasesWithWorkflowId(PK, workflowId) {
+  const params = {
+    TableName: config.cases.tableName,
+    KeyConditionExpression: 'PK = :pk',
+    FilterExpression: 'details.workflowId = :workflowId',
+    ExpressionAttributeValues: {
+      ':pk': PK,
+      ':workflowId': workflowId,
+    },
+  };
+
+  const [dynamoQueryError, queryCasesResult] = await to(dynamoDB.call('query', params));
+  if (dynamoQueryError) {
+    throw dynamoQueryError;
+  }
+
+  return queryCasesResult.Items;
+}
+
+async function putRecurringVivaCase(PK, workflowId, period) {
```

Comment:
Kanske kan förenklas så att vi får färre antal argument. Lär gärna på sidan 40 i boken Clean Code för en utförligt förklaring.

Reply:
Vad jag skulle kunna göra är att lägga till ett details argument dvs putRecurringVivaCase(PK, vivaDetails). Men på något sätt behöver jag få in workflowId och period i Item objektet i funktionen.

Reply:
Låt oss resonera....